### PR TITLE
Add metrics data and metric descriptor, to avoid using proto

### PR DESCRIPTION
**Pull request by bogdandrutu**
After this PR will add the builders for all metric types.

### CODE REVIEW

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/MetricData.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+package io.opentelemetry.sdk.metrics;
+
+import com.google.auto.value.AutoValue;
+import io.opentelemetry.trace.Timestamp;
+import javax.annotation.concurrent.Immutable;
+
+@Immutable
+@AutoValue
+public abstract class MetricData {
```
Comment (songy23) [inline sdk/src/main/java/io/opentelemetry/sdk/metrics/MetricData.java]: Nit: Javadoc.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/MetricDescriptor.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.concurrent.Immutable;
+
+@Immutable
+@AutoValue
+public abstract class MetricDescriptor {
```
Comment (songy23) [inline sdk/src/main/java/io/opentelemetry/sdk/metrics/MetricDescriptor.java]: ditto

File: sdk/src/test/java/io/opentelemetry/sdk/metrics/MetricDataTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import static com.google.common.truth.Truth.assertThat;
+
+import com.google.common.testing.EqualsTester;
+import io.opentelemetry.trace.Timestamp;
+import java.util.Collections;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class MetricDataTest {
```
Comment (songy23) [inline sdk/src/test/java/io/opentelemetry/sdk/metrics/MetricDataTest.java]: Missing Javadoc and `Runwith`.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/MetricData.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+   */
+  public abstract MetricDescriptor getMetricDescriptor();
+
+  /**
+   * Returns the type of this metric.
+   *
+   * @return the type of this metric.
+   * @since 0.1.0
+   */
+  public abstract Type getType();
```
Comment (songy23) [inline sdk/src/main/java/io/opentelemetry/sdk/metrics/MetricData.java]: Why type is not part of descriptor?

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/MetricData.java (hunk 1 above)
Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/MetricData.java]: Done.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/MetricDescriptor.java (hunk 2 above)
Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/MetricDescriptor.java]: Done.

File: sdk/src/test/java/io/opentelemetry/sdk/metrics/MetricDataTest.java (hunk 3 above)
Comment (bogdandrutu) [reply sdk/src/test/java/io/opentelemetry/sdk/metrics/MetricDataTest.java]: Done.

File: sdk/src/main/java/io/opentelemetry/sdk/metrics/MetricData.java (hunk 4 above)
Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/metrics/MetricData.java]: Thought more, makes sense to have it in the descriptor.

**Comment by bogdandrutu** [comment]
@songy23 PTAL

File: exporters/jaeger/src/main/java/io/opentelemetry/exporters/jaeger/JaegerGrpcSpanExporter.java [hunk 5]
```diff
@@ -32,7 +32,7 @@
 import java.util.logging.Logger;
 import javax.annotation.concurrent.ThreadSafe;
 
-/** Exports spans to Jaeger via gRPC, using Jaeger's protobuf model. */
+/** Exports spans to Jaeger via gRPC, using Jaeger's protobuf data. */
```
Comment (songy23) [inline exporters/jaeger/src/main/java/io/opentelemetry/exporters/jaeger/JaegerGrpcSpanExporter.java]: Is this change intentional?

Comment (bogdandrutu) [reply exporters/jaeger/src/main/java/io/opentelemetry/exporters/jaeger/JaegerGrpcSpanExporter.java]: reverted for the moment the change to move classes to data. this was clearly not intentional but I initially moved things to model then to data and idea thought is good to change model->data everywhere :)

**Comment by bogdandrutu** [comment]
@tylerbenson we do not expose in the API data format to avoid locking people to any specific format.

**Comment by sirianni** [comment]
What's the rationale for not using the `Metric` proto directly for the core data model?  IMO that's one of the confusing aspects of OpenCensus that I was happy to see changed in OpenTelemetry (@songy23 's original implementation PR #541)

**Comment by songy23** [comment]
I believe the discussion on POJOs vs. gen-proto classes originated from https://github.com/open-telemetry/opentelemetry-java/pull/543#issuecomment-532901254.

**Comment by sirianni** [comment]
Thanks @songy23 for the pointer.

There's nothing about the POJOs that require you to serialize them in protobuf format.  They can just as easily be serialized a JSON or simply used as in-memory data structures.  :thinking:

**Comment by songy23** [comment]
The main benefit I saw with using POJOs (SpanData, MetricData) is, for some internal fields we don't need extra conversions or object allocations. For example consider `TraceId`, `SpanId`, if we were to export `Span`s as protos, we have to first covert these fields to `ByteString`, whereas if we use `SpanData` we can simply reuse and expose the `TraceId`, `SpanId` classes.

**Comment by bogdandrutu** [comment]
@sirianni the other major benefit is to avoid extra dependencies like protobuf in our core SDK package.

**Comment by sirianni** [comment]
OK - this makes sense.  Especially the dependencies part.  Thanks for the quick replies @bogdandrutu and @songy23 .
